### PR TITLE
Generate: consistently handle special tokens as tensors

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

(reopened from #29788, requirements were merged and made this PR simpler)

To enable `torch.compile` with `generate`, some special token-related operations have to be rewritten into torch operations. That requires special tokens to be tensors instead of integers or a list of integers. (See https://github.com/huggingface/transformers/pull/29374 for a working prototype)

This PR reworks special token usage in `generate` to consistently treat them as a tensor, as opposed to e.g. keeping track of `eos_token_id` in integer and in tensor form.

👉 Review suggestion: start by reading `_prepare_special_tokens` and how it fits in `generate`.

______________________________________________________________________
Tests ran locally:
- [x] logits processors doctests (`pytest --doctest-modules src/transformers/generation/logits_process.py -vv`)
- [x] generate doctests (`pytest --doctest-modules src/transformers/generation/utils.py -vv`)
- [x] generate integration tests (`RUN_SLOW=1 py.test tests/generation/ -vv`)
- [x] cache integration tests (`RUN_SLOW=1 py.test tests/test_cache_utils.py -vv`)
- [x] llama slow tests (`RUN_SLOW=1 py.test tests/models/llama/test_modeling_llama.py -vv`)
- [x] whisper slow tests (`RUN_SLOW=1 py.test tests/models/whisper/test_modeling_whisper.py -vv`) -- same failures as in `main`

### CODE REVIEW

**Comment by gante** [comment]
cc @zucchini-nlp this one is the same as #29788, which you've already reviewed in its early state. Note that because we've merged other PRs first (e.g. removing the decoding functions from the public API), the diff is much smaller 💛

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -137,23 +137,26 @@ class MinLengthLogitsProcessor(LogitsProcessor):
...
-            eos_token_id = [eos_token_id]
-        if not all(isinstance(i, int) for i in eos_token_id) or any(i < 0 for i in eos_token_id):
-            logger.warning(f"`eos_token_id` has to be a list of positive integers, but is {eos_token_id}")
+        if not isinstance(eos_token_id, torch.Tensor):
+            if isinstance(eos_token_id, int):
+                eos_token_id = [eos_token_id]
+            eos_token_id = torch.tensor(eos_token_id)
+
+        if torch.is_floating_point(eos_token_id) or (eos_token_id < 0).any():
+            logger.warning_once(f"`eos_token_id` has to be a list of positive integers, but is {eos_token_id}")
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:150]: IMO we should just always cast to int.

Comment (ArthurZucker) [reply src/transformers/generation/logits_process.py:150]: Or leave it to error out on its own no?

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -137,23 +137,26 @@ class MinLengthLogitsProcessor(LogitsProcessor):
...
         self.min_length = min_length
         self.eos_token_id = eos_token_id
 
     @add_start_docstrings(LOGITS_PROCESSOR_INPUTS_DOCSTRING)
     def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
         vocab_tensor = torch.arange(scores.shape[-1], device=scores.device)
-        eos_token_id = torch.tensor(self.eos_token_id, device=scores.device)
-        eos_token_mask = torch.isin(vocab_tensor, eos_token_id)
+        self.eos_token_id = self.eos_token_id.to(scores.device)
+        eos_token_mask = torch.isin(vocab_tensor, self.eos_token_id)
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:159]: torch.isin not gonna work for mac but it's already there so no worries

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -1156,18 +1164,22 @@ class NoBadWordsLogitsProcessor(SequenceBiasLogitsProcessor):
...
-            filter(lambda bad_token_seq: all(bad_token_seq != [i] for i in eos_token_id), bad_words_ids)
-        )
+        if eos_token_id is not None:
+            if not isinstance(eos_token_id, torch.Tensor):
+                if isinstance(eos_token_id, int):
+                    eos_token_id = [eos_token_id]
+                eos_token_id = torch.tensor(eos_token_id)
+
+            bad_words_ids = list(
+                filter(lambda bad_token_seq: all(bad_token_seq != [i] for i in eos_token_id), bad_words_ids)
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1181]: this one does not need to be a tensor?

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -1471,15 +1482,22 @@ class ForcedEOSTokenLogitsProcessor(LogitsProcessor):
...
-            eos_token_id = [eos_token_id]
+
+        if not isinstance(eos_token_id, torch.Tensor):
+            if isinstance(eos_token_id, int):
+                eos_token_id = [eos_token_id]
+            eos_token_id = torch.tensor(eos_token_id)
         self.eos_token_id = eos_token_id
 
+        if torch.is_floating_point(eos_token_id) or (eos_token_id < 0).any():
+            logger.warning_once(f"`eos_token_id` has to be a list of positive integers, but is {eos_token_id}")
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1495]: again IMO erroring out would be best here, I don't know what will happen with negative / float points

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -448,20 +448,37 @@ def _maybe_initialize_input_ids_for_generation(
...
-            return torch.ones(inputs.shape[:2], dtype=torch.long, device=inputs.device)
+        if not is_input_ids:
+            return default_attention_mask
+
+        # Otherwise we have may have information -> try to infer the attention mask
+        if inputs.device.type == "mps":
+            # mps does not support torch.isin (https://github.com/pytorch/pytorch/issues/77764)
+            raise ValueError(
+                "Can't infer missing attention mask on `mps` device. Please provide an `attention_mask` or use a different device."
+            )
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:468]: Good

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -1334,6 +1327,58 @@ def _get_static_cache(self, max_batch_size: int, max_cache_len: int) -> StaticCa
...
 
+    def _prepare_special_tokens(
+        self, generation_config: GenerationConfig, kwargs_has_attention_mask: Optional[bool] = None
+    ) -> Tuple[Optional[torch.Tensor]]:
+        """
+        Prepares the special tokens for generation, overwriting the generation config with their processed versions
+        converted to tensor.
+
+        Note that `generation_config` is changed in place and stops being serializable after this method is called.
+        That is no problem is callen within `generate` (`generation_config` is a local copy that doesn't leave the
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1338]:
```suggestion
        That is no problem if called within `generate` (`generation_config` is a local copy that doesn't leave the
```

File: src/transformers/generation/utils.py [hunk 7]
```diff
@@ -1448,46 +1493,19 @@ def generate(
...
-        requires_attention_mask = "encoder_outputs" not in model_kwargs
-
-        if model_kwargs.get("attention_mask", None) is None and requires_attention_mask and accepts_attention_mask:
-            model_kwargs["attention_mask"] = self._prepare_attention_mask_for_generation(
-                inputs_tensor, generation_config.pad_token_id, generation_config.eos_token_id
-            )
-
-        # decoder-only models should use left-padding for generation
-        if not self.config.is_encoder_decoder:
+        # decoder-only models must use left-padding for generation.
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1507]: for compiled? Otherwise it shouldnot be a must no?

File: tests/models/seamless_m4t/test_modeling_seamless_m4t.py [hunk 8]
```diff
@@ -414,9 +414,11 @@ def _get_encoder_outputs(
         encoder_outputs["last_hidden_state"] = encoder_outputs.last_hidden_state.repeat_interleave(
             num_interleave, dim=0
         )
+        generation_config = copy.deepcopy(model.generation_config)
+        generation_config = model._prepare_special_tokens(generation_config)
```
Comment (ArthurZucker) [inline tests/models/seamless_m4t/test_modeling_seamless_m4t.py:418]: IMO feels a bit weird that preparing the special tokens returns a genertaion config. Would maybe not return and do inplace

File: tests/models/seamless_m4t/test_modeling_seamless_m4t.py [hunk 9]
```diff
@@ -414,9 +414,11 @@ def _get_encoder_outputs(
...
         input_ids = (
             torch.zeros(input_ids.shape[:2], dtype=torch.int64, layout=input_ids.layout, device=input_ids.device)
-            + model._get_decoder_start_token_id()
+            + generation_config.decoder_start_token_id
```
Comment (ArthurZucker) [inline tests/models/seamless_m4t/test_modeling_seamless_m4t.py:421]: that is very nice tho

File: src/transformers/generation/utils.py [hunk 10]
```diff
@@ -1334,6 +1327,58 @@ def _get_static_cache(self, max_batch_size: int, max_cache_len: int) -> StaticCa
...
+        """
+        Prepares the special tokens for generation, overwriting the generation config with their processed versions
+        converted to tensor.
+
+        Note that `generation_config` is changed in place and stops being serializable after this method is called.
+        That is no problem is callen within `generate` (`generation_config` is a local copy that doesn't leave the
+        function). However, if called outside `generate`, consider creating a copy of `generation_config` first.
+        """
+
+        # Convert special tokens to tensors (if they exist)
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1342]: It feels like it would be easier to do this when we load them in the generation config, and then when we save them we call `tolist()`? no? 

One thing is that `_tensor_or_none` seems to be used elsewhere a lot, even if it's potentially breaking, (deprecation cycle?) having tensor all the time should help no? This way there is no " preparation" for the decoder input ids if they are already saved in the correct format or something like that

**Review by ArthurZucker — COMMENTED** [review]
Overall good. Don't think we should warn but error out, and maybe update the serialize / de-serialize?

File: src/transformers/generation/logits_process.py (hunk 1 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:150]: in some cases it was a warning (bad practice), in others it was a deal breaker (would crash).

I've moved the bad practice warning to `_prepare_special_tokens` and changed the deal breaker cases to exceptions

File: src/transformers/generation/logits_process.py (hunk 2 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:159]: yeah, let's keep some pressure on the torch side 😈

File: src/transformers/generation/logits_process.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:1181]: `bad_words_ids`? it's actually good that it's an iterable, at least with the current implementation of the logits processor

File: src/transformers/generation/logits_process.py (hunk 4 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:1495]: (see above, changed to exception)

File: src/transformers/generation/utils.py (hunk 7 above)
Comment (gante) [reply src/transformers/generation/utils.py:1515]: added clarification: "generation" -> "batched generation"

File: src/transformers/generation/utils.py (hunk 10 above)
Comment (gante) [reply src/transformers/generation/utils.py:1350]: uhmmm let me give it a go, to see if something breaks

**Comment by gante** [comment]
@ArthurZucker addressed your comments :D

File: src/transformers/generation/logits_process.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/generation/logits_process.py:1175]: Yeah, but since we are converting everything to tensors I was wondering why this one is kept as. a list !

File: src/transformers/generation/utils.py (hunk 10 above)
Comment (ArthurZucker) [reply src/transformers/generation/utils.py:1350]: 😢 I see.

File: tests/models/seamless_m4t/test_modeling_seamless_m4t.py [hunk 11]
```diff
@@ -414,9 +414,11 @@ def _get_encoder_outputs(
         encoder_outputs["last_hidden_state"] = encoder_outputs.last_hidden_state.repeat_interleave(
             num_interleave, dim=0
         )
+        generation_config = copy.deepcopy(model.generation_config)
+        model._prepare_special_tokens(generation_config)
         input_ids = (
             torch.zeros(input_ids.shape[:2], dtype=torch.int64, layout=input_ids.layout, device=input_ids.device)
-            + model._get_decoder_start_token_id()
+            + generation_config.decoder_start_token_id
```
Comment (ArthurZucker) [inline tests/models/seamless_m4t/test_modeling_seamless_m4t.py:421]: do we have to do the + and not just use the decoder start toekn id?

**Review by ArthurZucker — APPROVED** [review]
Looks good thanks for updating

File: tests/models/seamless_m4t/test_modeling_seamless_m4t.py (hunk 11 above)
Comment (gante) [reply tests/models/seamless_m4t/test_modeling_seamless_m4t.py:421]: I think the + here makes it a tensor of all `decoder_start_token_id` (as opposed to concatenation).
